File: 项目/OC项目/微博(OC)/微博(OC)/Classes/Main/View/TextView.m
//
//  TextView.m
//  微博(OC)
//
//  Created by 金亮齐 on 2017/5/12.
//  Copyright © 2017年 醉看红尘这场梦. All rights reserved.
//
#import "TextView.h"

@implementation TextView

-(id)initWithFrame:(CGRect)frame {
    
    self = [super initWithFrame:frame];
    if (self) {
    
        // 当UITextView的文字发生改变时，UITextView自己会发出一个UITextViewTextDidChangeNotification通知

        [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(textDidChange) name:UITextViewTextDidChangeNotification object:self];
        
    }
    
    
    return self;
    
}



-(void)setPlaceholder:(NSString *)placeholder {
    
    _placeholder = [placeholder copy];
    
    [self setNeedsDisplay];
    
    
}


-(void)setPlaceholderColor:(UIColor *)placeholderColor {
    
    _placeholderColor = placeholderColor;
    [self setNeedsDisplay];
    
}

-(void)setText:(NSString *)text {
    
    [super setText:text];
    
    [self setNeedsDisplay];
    
}


- (void)setFont:(UIFont *)font
{
    [super setFont:font];
    
    [self setNeedsDisplay];
}

//移除通知
-(void)dealloc {
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    
}

-(void)setAttributedText:(NSAttributedString *)attributedText {
    
    [super setAttributedText:attributedText];
    [self setNeedsDisplay];
    
}

/**
 * 监听文字改变
 */
-(void)textDidChange {
    
    //重绘
    [self setNeedsDisplay];
    
}

-(void)drawRect:(CGRect)rect {
    
    if (self.hasText) return;
    
    //文字属性
    NSMutableDictionary *attr = [NSMutableDictionary dictionary];
    attr[NSFontAttributeName] = self.font;
    attr[NSForegroundColorAttributeName] = self.placeholderColor?self.placeholderColor:[UIColor grayColor];
    
    // 画文字
    CGFloat x = 5;
    CGFloat w = rect.size.width - 2 * x;
    CGFloat y = 8;
    CGFloat h = rect.size.height - 2 * y;
    CGRect placeholderRect = CGRectMake(x, y, w, h);
    [self.placeholder drawInRect:placeholderRect withAttributes:attr];
    

}

@end

File: 项目/Swift项目/百思不得姐/百思不得姐/Classes/Publish(公共)/View/TextView.swift
//
//  TextView.swift
//  百思不得姐
//
//  Created by 金亮齐 on 2017/11/13.
//  Copyright © 2017年 金亮齐. All rights reserved.
//

import UIKit

class TextView: UITextView {

    var placeholder: String = ""
    
    override init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
        let name = NSNotification.Name.UITextViewTextDidChange
        NotificationCenter.default.addObserver(self, selector: #selector(textViewChange), name: name, object: nil)
    }
    
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func draw(_ rect: CGRect) {
        
        if !self.text.isEmpty {
            return
        }
        let x: CGFloat = 3.0
        let y: CGFloat = 7.0
        let textRect = CGRect(x: x, y: y, width: screenWidth - 2 * x, height: frame.height)
        let text = placeholder as NSString
        text.draw(in: textRect, withAttributes: [NSAttributedStringKey.font: UIFont.systemFont(ofSize: 15), NSAttributedStringKey.foregroundColor : UIColor.lightGray])
    }
    
    @objc private func textViewChange() {
        setNeedsDisplay()
    }

}
